src/main/java/App.java

```diff
@@ -6,16 +7,38 @@
 import static spark.Spark.*;
 
 public class App {
-    private static long counter = 0;
+    private static ConcurrentHashMap<String, Integer> map = new ConcurrentHashMap<String, Integer>(1);
 
     public static void main(String[] args) {
         staticFileLocation("public");
 
         get("/", (req, res) -> {
+            // Intentionally wrote a poor code to test Amazon CodeGuru
+            int counter;
+            if (map.containsKey("counter")) {
+                counter = map.get("counter");
+            } else {
+                map.put("counter", 0);
+                counter = 0;
+            }
+
             return new ModelAndView(ImmutableMap.of("counter", counter), "index.mustache");
         }, new MustacheTemplateEngine());
 
-        post("/roar", (req, res) -> String.format("%d", ++counter));
+        post("/roar", (req, res) -> {
+            // Intentionally wrote a poor code to test Amazon CodeGuru
+            int counter;
+            if (map.containsKey("counter")) {
+                counter = map.get("counter");
+                ++counter;
+                map.put("counter", counter);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **31 and 34**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **40**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.